smarts/core/plan.py

```diff
@@ -0,0 +1,337 @@
+# Copyright (C) 2020. Huawei Technologies Co., Ltd. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+# to allow for typing to refer to class being defined (Mission)...
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from functools import lru_cache
+import math
+import random
+from typing import List, Optional, Tuple
+
+import numpy as np
+
+from .coordinates import Dimensions, Heading, Point, Pose, RefLinePoint
+from .road_map import RoadMap
+
+from smarts.sstudio.types import EntryTactic, TrapEntryTactic
+
+
+class PlanningError(Exception):
+    pass
+
+
+# XXX: consider using smarts.core.coordinates.Pose for this
+@dataclass(frozen=True)
+class Start:
+    position: Tuple[int, int]
+    heading: Heading
+    from_front_bumper: Optional[bool] = True
+
+    @property
+    def point(self) -> Point:
+        return Point(*self.position)
+
+    @classmethod
+    def from_pose(cls, pose: Pose):
+        return cls(
+            position=pose.position[:2],
+            heading=pose.heading,
+            from_front_bumper=False,
+        )
+
+
+@dataclass(frozen=True)
+class Goal:
+    def is_endless(self) -> bool:
+        return True
+
+    def is_reached(self, vehicle) -> bool:
+        return False
```

Comment:
This seems curious to me that `Goal` is defined the same way as `EndlessGoal`. My preference, if possible, is that this base is abstract.

Reply:
This was how it was in the original (just moved here from `scenario.py` where it was before).   Although I didn't write it myself, I suspect the intention here was to define the "interface" for Goals, i.e., methods that all derived classes must support, for purposes of polymorphism.  If these methods are moved down into `EndlessGoal`, then an empty `Goal` as a base class serves very little purpose.

Reply:
Yes, sorry, I realise that this is pre-existing. My thought here is that derived classes are not required to implement the interface because the methods (i.e. `is_endless()`, `is_reached()`) have already been implemented at the `Goal` level. This makes the empty `EndlessGoal` redundant except for the type. And then `PositionalGoal` is not required to implement those methods.